README.md
# NAME

Badge::Depot::Plugin::Cpantesters - CPAN testers plugin for Badge::Depot

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.10+-blue.svg" alt="Requires Perl 5.10+" />
    <a href="https://travis-ci.org/Csson/p5-Badge-Depot-Plugin-CpanTesters"><img src="https://api.travis-ci.org/Csson/p5-Badge-Depot-Plugin-CpanTesters.svg?branch=master" alt="Travis status" /></a>
    <a href="http://cpants.cpanauthors.org/dist/Badge-Depot-Plugin-Cpantesters-0.0102"><img src="http://badgedepot.code301.com/badge/kwalitee/CSSON/Badge-Depot-Plugin-Cpantesters/0.0102" alt="Distribution kwalitee" /></a>
    <a href="http://matrix.cpantesters.org/?dist=Badge-Depot-Plugin-Cpantesters%200.0102"><img src="http://badgedepot.code301.com/badge/cpantesters/Badge-Depot-Plugin-Cpantesters/0.0102" alt="CPAN Testers result" /></a>
    <img src="https://img.shields.io/badge/coverage-70.6%-red.svg" alt="coverage 70.6%" />
    </p>
</div>

# VERSION

Version 0.0102, released 2016-08-11.

# SYNOPSIS

If used standalone:

    use Badge::Depot::Plugin::Cpantesters;

    my $badge = Badge::Depot::Plugin::Cpantesters->new(dist => 'The-Dist', version => '0.1002');

    print $badge->to_html;
    # prints:
    <a href="http://matrix.cpantesters.org/?dist=The-Dist%200.1002">
        <img src="http://badgedepot.code301.com/badge/cpantesters/The-Dist/0.1002" alt="CPAN Testers result" />
    </a>

If used with [Pod::Weaver::Section::Badges](https://metacpan.org/pod/Pod::Weaver::Section::Badges), in weaver.ini:

    [Badges]
    ; other settings
    badge = cpantesters

# DESCRIPTION

Creates a [CpanTesters](http://cpantesters.org) badge for a distribution.

This class consumes the [Badge::Depot](https://metacpan.org/pod/Badge::Depot) role.

# ATTRIBUTES

This badge tries to use distribution meta data to set the attributes. If that is available no attributes need to be set manually. The following checks are made:

1. If the badge is used via [Pod::Weaver::Section::Badges](https://metacpan.org/pod/Pod::Weaver::Section::Badges) during a [Dist::Zilla](https://metacpan.org/pod/Dist::Zilla) build, then `version` and `dist` are set to the values in the Dist::Zilla object.
2. If there is a `META.json` in the distribution root then that is used to set `version` and `dist`.

If neither of those are true then `dist` and `version` should passed to the constructor.

## dist

Distribution name. With dashes, not colons.

## version

Distribution version.

## base\_url

Default: `https://badgedepot.code301.com`

Set this if you wish to use another instance of [Badge::Depot::App](https://metacpan.org/pod/Badge::Depot::App).

# SEE ALSO

- [Badge::Depot](https://metacpan.org/pod/Badge::Depot)
- [Task::Badge::Depot](https://metacpan.org/pod/Task::Badge::Depot)

# SOURCE

[https://github.com/Csson/p5-Badge-Depot-Plugin-CpanTesters](https://github.com/Csson/p5-Badge-Depot-Plugin-CpanTesters)

# HOMEPAGE

[https://metacpan.org/release/Badge-Depot-Plugin-Cpantesters](https://metacpan.org/release/Badge-Depot-Plugin-Cpantesters)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Badge-Depot-Plugin-Cpantesters

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.0102 2016-08-11T23:33:25Z UTC
 [Other Changes]
 - Default url protocol (temporarily) changed from https to http.

 [Stats & Dependency Changes]
 - 2 files changed, 4 insertions(+), 4 deletions(-)
 - (dev req) + Dist::Zilla::Plugin::MetaConfig (any)
 - (dev req) + Dist::Zilla::Plugin::PromptIfStale (any)
 - (dev sug) ~ Dist::Iller 0.1407 --> 0.1408
 - (dev sug) ~ Dist::Iller::Config::Author::CSSON 0.0309 --> 0.0311

0.0101 2016-02-22T15:04:30Z UTC
 [Enhancements]
 - Now uses the Dist::Zilla object (if it is used during a Dist::Zilla
   build) to set the 'dist' and 'version' attributes.

 [Stats & Dependency Changes]
 - 2 files changed, 31 insertions(+), 9 deletions(-)
 - (run req) ~ Badge::Depot 0.0103 --> 0.0104
 - (run req) ~ JSON::MaybeXS (any) --> 1.003003
 - (run req) ~ MooseX::AttributeShortcuts (any) --> 0.028
 - (run req) ~ Path::Tiny (any) --> 0.072
 - (dev req) + Badge::Depot 0.0103
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Cpantesters (any)
 - (dev req) + Badge::Depot::Plugin::Kwalitee (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Badge::Depot::Plugin::Travis (any)
 - (dev req) + Dist::Zilla 5.000
 - (dev req) + Dist::Zilla::Plugin::Git 2.036
 - (dev req) + Pod::Weaver::Section::Badges 0.0401
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::Version 1
 - (dev req) ~ Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   (any) --> 0.007
 - (dev req) ~ Dist::Zilla::Plugin::ChangeStats::Dependencies::Git (any)
   --> 0.0100
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::Clean (any) --> 0.07
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.151680
 - (dev req) ~ Dist::Zilla::Plugin::Git::Contributors (any) --> 0.017
 - (dev req) ~ Dist::Zilla::Plugin::GithubMeta (any) --> 0.54
 - (dev req) ~ Dist::Zilla::Plugin::InstallRelease (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.003001
 - (dev req) ~ Dist::Zilla::Plugin::NextRelease::Grouped (any) --> 0.0200
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod (any) --> 0.150250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests (any) --> 0.028
 - (dev req) ~ Dist::Zilla::Plugin::Test::Compile (any) --> 2.054
 - (dev req) ~ Dist::Zilla::Plugin::Test::EOF (any) --> 0.0501
 - (dev req) ~ Dist::Zilla::Plugin::Test::EOL (any) --> 0.18
 - (dev req) ~ Dist::Zilla::Plugin::Test::Kwalitee::Extra (any) --> v0.2.1
 - (dev req) ~ Dist::Zilla::Plugin::Test::NoTabs (any) --> 0.15
 - (dev req) ~ Dist::Zilla::Plugin::Test::ReportPrereqs (any) --> 0.021
 - (dev req) ~ Dist::Zilla::Plugin::Test::Version (any) --> 1.05
 - (dev req) ~ Dist::Zilla::Plugin::TravisYML (any) --> 1.15
 - (dev req) ~ Test::NoTabs (any) --> 1.4
 - (dev req) ~ Test::Warnings (any) --> 0.026
 - (dev req) - Dist::Zilla::Plugin::MetaProvides::Class
 - (dev req) - Dist::Zilla::Plugin::PodnameFromClassname
 - (dev req) - Dist::Zilla::Plugin::Prereqs::Plugins
 - (dev req) - Pod::Weaver::Section::Homepage::DefaultCPAN
 - (dev req) - Pod::Weaver::Section::Source::DefaultGitHub
 - (dev sug) ~ Dist::Iller 0.1406 --> 0.1407
 - (dev sug) ~ Dist::Iller::Config::Author::CSSON 0.0307 --> 0.0309

0.0100 2016-02-17T15:59:34Z UTC
 - Initial release

 [Stats & Dependency Changes]
 - (run req) + Badge::Depot 0.0103
 - (run req) + JSON::MaybeXS (any)
 - (run req) + Moose 2.1400
 - (run req) + MooseX::AttributeShortcuts (any)
 - (run req) + Path::Tiny (any)
 - (run req) + Types::Standard 1.000000
 - (run req) + Types::URI (any)
 - (run req) + namespace::autoclean 0.18
 - (run req) + perl 5.010000
 - (test req) + ExtUtils::MakeMaker (any)
 - (test req) + File::Spec (any)
 - (test req) + IO::Handle (any)
 - (test req) + IPC::Open3 (any)
 - (test req) + Test::More 0.96
 - (test rec) + CPAN::Meta 2.120900
 - (conf req) + ExtUtils::MakeMaker (any)
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   (any)
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git (any)
 - (dev req) + Dist::Zilla::Plugin::CheckChangesHasContent (any)
 - (dev req) + Dist::Zilla::Plugin::Clean (any)
 - (dev req) + Dist::Zilla::Plugin::ConfirmRelease (any)
 - (dev req) + Dist::Zilla::Plugin::CopyFilesFromBuild (any)
 - (dev req) + Dist::Zilla::Plugin::DistIller::MetaGeneratedBy (any)
 - (dev req) + Dist::Zilla::Plugin::ExecDir (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Check (any)
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Contributors (any)
 - (dev req) + Dist::Zilla::Plugin::Git::GatherDir (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Push (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Tag (any)
 - (dev req) + Dist::Zilla::Plugin::GithubMeta (any)
 - (dev req) + Dist::Zilla::Plugin::InstallRelease (any)
 - (dev req) + Dist::Zilla::Plugin::License (any)
 - (dev req) + Dist::Zilla::Plugin::MakeMaker (any)
 - (dev req) + Dist::Zilla::Plugin::Manifest (any)
 - (dev req) + Dist::Zilla::Plugin::ManifestSkip (any)
 - (dev req) + Dist::Zilla::Plugin::MetaJSON (any)
 - (dev req) + Dist::Zilla::Plugin::MetaNoIndex (any)
 - (dev req) + Dist::Zilla::Plugin::MetaProvides::Class (any)
 - (dev req) + Dist::Zilla::Plugin::MetaProvides::Package (any)
 - (dev req) + Dist::Zilla::Plugin::MetaYAML (any)
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped (any)
 - (dev req) + Dist::Zilla::Plugin::PodSyntaxTests (any)
 - (dev req) + Dist::Zilla::Plugin::PodWeaver (any)

lib/Badge/Depot/Plugin/CpanTesters.pm
use 5.10.0;
use strict;
use warnings;

package Badge::Depot::Plugin::CpanTesters;

# ABSTRACT: CPAN testers plugin for Badge::Depot
# AUTHORITY
our $VERSION = '0.0100';

use Moose;
use namespace::autoclean;
use MooseX::AttributeShortcuts;
use Types::Standard qw/Str HashRef/;
use Types::URI qw/Uri/;
use JSON::MaybeXS 'decode_json';
use Path::Tiny;
with 'Badge::Depot';

has dist => (
    is => 'ro',
    isa => Str,
    lazy => 1,
    default => sub {
        my $self = shift;
        if($self->has_meta) {
            return $self->_meta->{'dist'} if exists $self->_meta->{'dist'};
        }
    },
);
has version => (
    is => 'ro',
    isa => Str,
    lazy => 1,
    default => sub {
        my $self = shift;
        if($self->has_meta) {
            return $self->_meta->{'version'} if exists $self->_meta->{'version'};
        }
    },
);
has base_url => (
    is => 'ro',
    isa => Uri,
    coerce => 1,
    lazy => 1,
    default => 'https://badgedepot.code301.com',
);
has custom_image_url => (
    is => 'ro',
    isa => Uri,
    coerce => 1,
    lazy => 1,
    builder => 1,
);

sub _build_custom_image_url  {
    my $self = shift;
    return sprintf '%s/badge/cpantesters/%s/%s', $self->base_url, $self->dist, $self->version;
}
has _meta => (
    is => 'ro',
    isa => HashRef,
    predicate => 'has_meta',
    builder => '_build_meta',
);

sub _build_meta {
    my $self = shift;

    return {} if !path('META.json')->exists;

    my $json = path('META.json')->slurp_utf8;
    my $data = decode_json($json);

    return {} if !exists $data->{'name'} || !exists $data->{'version'};

    return {
        dist => $data->{'name'},
        version => $data->{'version'},
    };
}

sub BUILD {
    my $self = shift;
    $self->link_url(sprintf 'http://matrix.cpantesters.org/?dist=%s %s', $self->dist, $self->version eq 'latest' ? '' : $self->version);
    $self->image_url($self->custom_image_url);
    $self->image_alt('CPAN Testers result');
}

__PACKAGE__->meta->make_immutable;

1;

__END__

=pod

=head1 SYNOPSIS

    use Badge::Depot::Plugin::CpanTesters;

=head1 DESCRIPTION

Badge::Depot::Plugin::CpanTesters is ...

=head1 SEE ALSO

=cut


lib/Badge/Depot/Plugin/Cpantesters.pm
use 5.10.0;
use strict;
use warnings;

package Badge::Depot::Plugin::Cpantesters;

# ABSTRACT: CPAN testers plugin for Badge::Depot
# AUTHORITY
our $VERSION = '0.0103';

use Moose;
use namespace::autoclean;
use MooseX::AttributeShortcuts;
use Types::Standard qw/Str HashRef/;
use Types::URI qw/Uri/;
use JSON::MaybeXS 'decode_json';
use Path::Tiny;
with 'Badge::Depot';

has dist => (
    is => 'ro',
    isa => Str,
    lazy => 1,
    default => sub {
        my $self = shift;
        if($self->get_meta('dist')) {
            return $self->_meta->{'dist'};
        }
    },
);
has version => (
    is => 'ro',
    isa => Str,
    lazy => 1,
    default => sub {
        my $self = shift;
        if($self->get_meta('version')) {
            return $self->_meta->{'version'};
        }
    },
);
has base_url => (
    is => 'ro',
    isa => Uri,
    coerce => 1,
    lazy => 1,
    default => 'http://badgedepot.code301.com',
);
has custom_image_url => (
    is => 'ro',
    isa => Uri,
    coerce => 1,
    lazy => 1,
    builder => 1,
);

sub _build_custom_image_url  {
    my $self = shift;
    return sprintf '%s/badge/cpantesters/%s/%s', $self->base_url, $self->dist, $self->version;
}
has _meta => (
    is => 'ro',
    isa => HashRef,
    traits => ['Hash'],
    lazy => 1,
    predicate => 'has_meta',
    builder => '_build_meta',
    handles => {
        get_meta => 'get',
    },
);

sub _build_meta {
    my $self = shift;

    if($self->has_zilla) {
        return {
            dist => $self->zilla->name,
            version => $self->zilla->version,
        };
    }

    return {} if !path('META.json')->exists;

    my $json = path('META.json')->slurp_utf8;
    my $data = decode_json($json);

    return {} if !exists $data->{'name'} || !exists $data->{'version'};

    return {
        dist => $data->{'name'},
        version => $data->{'version'},
    };
}

sub BUILD {
    my $self = shift;
    $self->link_url(sprintf 'http://matrix.cpantesters.org/?dist=%s %s', $self->dist, $self->version eq 'latest' ? '' : $self->version);
    $self->image_url($self->custom_image_url);
    $self->image_alt('CPAN Testers result');
}

__PACKAGE__->meta->make_immutable;

1;

__END__

=pod

=head1 SYNOPSIS

If used standalone:

    use Badge::Depot::Plugin::Cpantesters;

    my $badge = Badge::Depot::Plugin::Cpantesters->new(dist => 'The-Dist', version => '0.1002');

    print $badge->to_html;
    # prints:
    <a href="http://matrix.cpantesters.org/?dist=The-Dist%200.1002">
        <img src="http://badgedepot.code301.com/badge/cpantesters/The-Dist/0.1002" alt="CPAN Testers result" />
    </a>

If used with L<Pod::Weaver::Section::Badges>, in weaver.ini:

    [Badges]
    ; other settings
    badge = cpantesters

=head1 DESCRIPTION

Creates a L<CpanTesters|http://cpantesters.org> badge for a distribution.

This class consumes the L<Badge::Depot> role.

=head1 ATTRIBUTES

This badge tries to use distribution meta data to set the attributes. If that is available no attributes need to be set manually. The following checks are made:

=for :list
1. If the badge is used via L<Pod::Weaver::Section::Badges> during a L<Dist::Zilla> build, then C<version> and C<dist> are set to the values in the Dist::Zilla object.
2. If there is a C<META.json> in the distribution root then that is used to set C<version> and C<dist>.

If neither of those are true then C<dist> and C<version> should passed to the constructor.

=for :list

=head2 dist

Distribution name. With dashes, not colons.

=head2 version

Distribution version.

=head2 base_url

Default: C<https://badgedepot.code301.com>

Set this if you wish to use another instance of L<Badge::Depot::App>.

=head1 SEE ALSO

=for :list
* L<Badge::Depot>
* L<Task::Badge::Depot>

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.007.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "CPAN testers plugin for Badge::Depot",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Badge-Depot-Plugin-Cpantesters",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "Badge::Depot::Plugin::Cpantesters",
  "PREREQ_PM" => {
    "Badge::Depot" => "0.0104",
    "JSON::MaybeXS" => "1.003003",
    "Moose" => "2.1400",
    "MooseX::AttributeShortcuts" => "0.028",
    "Path::Tiny" => "0.072",
    "Types::Standard" => "1.000000",
    "Types::URI" => 0,
    "namespace::autoclean" => "0.18"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.0103",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Badge::Depot" => "0.0104",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "JSON::MaybeXS" => "1.003003",
  "Moose" => "2.1400",
  "MooseX::AttributeShortcuts" => "0.028",
  "Path::Tiny" => "0.072",
  "Test::More" => "0.96",
  "Types::Standard" => "1.000000",
  "Types::URI" => 0,
  "namespace::autoclean" => "0.18"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
